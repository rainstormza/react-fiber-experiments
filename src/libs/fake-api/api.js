import fixtures from "./fixtures.json";

let db;

const DEFAULT_MODE = "fastNetwork";

// from general to specific
const defaultModes = {
  slowNetwork: {
    "/*": 4000
  },
  [DEFAULT_MODE]: {
    "/*": 150
  },
  slowEndPoint: {
    "/*": 2000,
    "/course/*": 4000
    // "/course/*/nextLesson": 6000
  }
};

/**
 * Transforms a config to a usable structure (see unit test for example)
 */
export const _configModes = (modes = defaultModes) => {
  return Object.entries(modes).reduce((acc, [mode, value]) => {
    const innerConfig = Object.entries(value).reduce(
      (innerAcc, [pattern, delay]) => {
        innerAcc.push({
          pattern: new RegExp(pattern.replace("*", "(\\.*)")),
          delay
        });
        return innerAcc;
      },
      []
    );
    acc[mode] = innerConfig.reverse();
    return acc;
  }, {});
};

export const listModes = () => Object.keys(defaultModes);

export const buildDb = (dump = fixtures) => {
  db = dump;
};

const LOCAL_STORAGE_MODE_KEY = "fake-api-network-mode";

export const saveNetworkMode = mode => {
  if (localStorage && localStorage.setItem) {
    return localStorage.setItem(LOCAL_STORAGE_MODE_KEY, mode);
  }
  console.warn(
    "No localStorage available, fallback in memory mode - reload based / state based actions may use default delay."
  );
};

export const getNetworkMode = () => {
  if (localStorage && localStorage.getItem) {
    const mode = localStorage.getItem(LOCAL_STORAGE_MODE_KEY);
    return mode || DEFAULT_MODE;
  }
  return DEFAULT_MODE;
};

const matchUrlToResponse = (url, data = db) => {
  // /courses - remove lessons
  if (url.startsWith("/courses")) {
    return data.reduce((acc, current) => {
      const { lessons, ...rest } = current;
      acc.push(rest);
      return acc;
    }, []);
  }
  // /course/${topic} - match topic
  if (url.startsWith("/course/") && !url.endsWith("/nextLesson")) {
    const courseMatch = url.match(/\/course\/(.*)/);
    if (courseMatch && courseMatch[1]) {
      const res = data.filter(course => course.id === courseMatch[1]);
      if (res && res[0]) {
        return res[0];
      }
    }
  }
  // /course/${topic}/nextLesson - pick a lesson
  return null;
};

const matchUrlToDelay = (url, mode, modes = defaultModes) => {
  const config = _configModes(modes)[mode];
  for (let i = 0; i < config.length; i++) {
    if (url.match(config[i].pattern)) {
      return config[i].delay;
    }
  }
  return 0;
};

export const makeFakeApi = (mode = getNetworkMode()) => url => {
  const response = matchUrlToResponse(url);
  const delay = matchUrlToDelay(url, mode);
  console.log(`[FAKE ${delay}ms] ${url}`);
  return new Promise(resolve => {
    setTimeout(() => {
      console.log(`[FAKE ${delay}ms] ${url}`, response);
      return resolve(response);
    }, delay);
  });
};
