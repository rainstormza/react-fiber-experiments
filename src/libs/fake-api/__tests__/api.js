import { _configModes } from "../api";

describe("/libs/fake-api", () => {
  describe("api", () => {
    describe("_configModes", () => {
      it("should prepare a ready to be used config by default", () => {
        const modes = _configModes();
        expect(modes).toMatchSnapshot();
      });
    });
  });
});
