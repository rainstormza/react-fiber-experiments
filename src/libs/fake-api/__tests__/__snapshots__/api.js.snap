// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/libs/fake-api api _configModes should prepare a ready to be used config by default 1`] = `
Object {
  "fastNetwork": Array [
    Object {
      "delay": 150,
      "pattern": /\\\\/\\(\\\\\\.\\*\\)/,
    },
  ],
  "slowEndPoint": Array [
    Object {
      "delay": 4000,
      "pattern": /\\\\/course\\\\/\\(\\\\\\.\\*\\)/,
    },
    Object {
      "delay": 2000,
      "pattern": /\\\\/\\(\\\\\\.\\*\\)/,
    },
  ],
  "slowNetwork": Array [
    Object {
      "delay": 4000,
      "pattern": /\\\\/\\(\\\\\\.\\*\\)/,
    },
  ],
}
`;
